---
title: "2021-OxsR-RNAseq-analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# This analysis identifies the differentially expressed genes across the genotype (H26 and oxsR) and NaOCl stress (absent and present).

### Data processing for the Single-end sequencing in Terminal or Computer cluster
```{r}
# Unzip raw data
#   $ gunzip *.gz

# Sequencing quality check and trimming adapter or low quality reads
#   $ fastqc -o fastqc *.fastq
#   $ trim_galore $FILENAME -o ./Trim_done *.fastq

# Alignment of the sequence reads to the index genome
#   $ bowtie2-build ~/GCF_000025685.1_ASM2568v1_genomic.fa HVO (Download genome from https://www.ncbi.nlm.nih.gov/assembly/?term=Haloferax+volcanii+DS2+genome)
#   $ bowtie2 -x ./Index_genome/HVO -U ./$FILENAME -S ./$FILENAME.sam

# File format change
# 1)  $ samtools view -bS $FILENAME -o ./bam/$FILENAME.bam
# 2)  $ samtools sort $FILENAME -o ./bam_sort/$FILENAME.sort.bam
# 3)  $ samtools index $FILENAME

# Read counts 
# htseq-count -t gene -i Name --stranded=no -f bam -r pos ./sorted_bamfile/$FILENAME ./reference/GCF_000025685.1_ASM2568v1_genomic.gff > ./count/$FILENAME.count

# RNA-seq data analysis
#Duke workshop: 2020 Quantitative Methods Workshops for HIV Researchers (https://ipyn-az-07.oit.duke.edu:51044/tree/hiv-workshop-2019-2020)
#DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#htseq
```

### Transfer the raw count to the normalized count by DESeq2
```{r}

rm(list = ls())  # Remove all the previous definitions 

library(tidyverse)
library(ggplot2)
library(DESeq2)
library(ggrepel)
library(RColorBrewer)
library(pheatmap)
library(dendextend)
library(EnhancedVolcano)
library(magrittr)

# Create a DESeqDataSet object
# Read the raw count file containing all bio reps
mycount <- as.matrix(read.csv("Count.csv", row.names="gene_id"))
dim(mycount)

# Read the meta file 
coldata <- read.csv("Meta.csv", row.names = 1)
#ggplot(coldata, aes(x = coldata$Label, y = coldata$Total_count)) + geom_point() + theme(axis.text.x = element_text(face="bold", color="blue", angle=45))
dim(coldata)
coldata <- coldata[,c("Genotype","Stress")]

# Are they identical as they follow different orders?
all(rownames(coldata) == colnames(mycount))
# Reorder the columns of the count matrix according to the order of Label in columnData
mycount <- mycount[,rownames(coldata)]
all(rownames(coldata) == colnames(mycount))

# Make DESeq object on the basis of the counts
dds.mydata <- DESeqDataSetFromMatrix(mycount, coldata,
                                  ~ Genotype + Stress + Genotype:Stress)  # "Stress:Genotype" means the effect of Stress on the Genotype effect (the difference of difference)

# Assign the comparison references
dds.mydata$Stress <- relevel(dds.mydata$Stress, ref = "No")
dds.mydata$Genotype <- relevel(dds.mydata$Genotype, ref = "H26")

dds.mydata@design # check your design for the comparison.
dds.mydata@colData # check your column data shows all the factors 

# Estimate Size Factors to inspect the raw count data which is generated from various lanes and bio reps. 
dds.mydata <- estimateSizeFactors(dds.mydata)
sizeFactors(dds.mydata)
mydf <- sizeFactors(dds.mydata) %>%
  as.data.frame %>%
  rownames_to_column
colnames(mydf)[2] <- "sizefac"
ggplot(mydf, aes(rowname, sizefac)) + geom_point() + theme(axis.text.x = element_text(face="bold", color="blue", angle=45))

# Convert the raw count to the normalized count and visualize the raw and normalized data
# Ref: http://pedagogix-tagc.univ-mrs.fr/courses/ASG1/practicals/rnaseq_diff_Snf2/rnaseq_diff_Snf2.html
library(affy)
#"mycount" is the raw data
#hist(mycount, col="blue", border="white", breaks=100)
#hist(log2(mycount+1), col="blue", border="white", breaks=100, main="Log2-transformed counts per gene with raw data", ylab="Number of genes") #histogram style
#boxplot(log2(mycount+1), horizontal=T) #boxplot style
plotDensity(log2(mycount+1), lty=1, lwd=2, ylab="Desity of the number of genes", xlab="Log2(Raw count + 1)")
#grid()

#normalized data
normalized_counts <- counts(dds.mydata, normalized=TRUE)
#hist(normalized_counts, col="blue", border="white", breaks=100)
#hist(log2(normalized_counts+1), col="blue", border="white", breaks=100, main="Log2-transformed counts per gene with normalized data", ylab="Number of genes")
plotDensity(log2(normalized_counts+1), lty=1, lwd=2, ylab="Desity of the number of genes", xlab="Log2(Normalized count + 1)")
#grid()

write.csv(normalized_counts, file="count_norm.csv")
```


### Differential Expression Analysis
```{r}
# We can now conduct a differential expression analysis using the DESeq() function. 
# Keep in mind that to get to this step, we first estimated the size factors and then the dispersion parameters.

ddsDE <- DESeq(dds.mydata)

resultsNames(ddsDE)
ddsDE # Look at object

# Total number of raw counts per sample
colSums(counts(ddsDE))

colSums(counts(ddsDE)) %>%
  as.data.frame %>%
  rownames_to_column -> mydf.raw.count

colnames(mydf.raw.count)[2] <- "whole.gene.count"
mydf.raw.count

ggplot(mydf.raw.count, aes(rowname, whole.gene.count)) + geom_point() + theme(axis.text.x = element_text(face="bold", color="blue", angle=45))

# Total number of normalized counts per sample
colSums(counts(ddsDE, normalized = T)) 

colSums(counts(ddsDE, normalized = T))  %>%
  as.data.frame %>%
  rownames_to_column -> mydf.norm.count

colnames(mydf.norm.count)[2] <- "whole.gene.norm.count"
mydf.norm.count

ggplot(mydf.norm.count, aes(rowname, whole.gene.norm.count)) + geom_point() + theme(axis.text.x = element_text(face="bold", color="blue", angle=45))

# We can get the results for the differential expression analysis.
ddsDE@design
resultsNames(ddsDE)
results(ddsDE)
```

### Clustering: Regularized log transformation.
```{r}
# The regularized log transform can be obtained using the rlog() function. Note that an important argument for this function is blind (TRUE by default). The default "blinds" the normalization to the design. This is very important so as to not bias the analyses (e.g. class discovery)
rld <- rlog(ddsDE, blind = TRUE)
rld.table <- assay(rld)
write.csv(rld.table, "count_norm_log2.csv") #edit the first column name as "Locus" in log2_norm_count.csv from your local folder
sampleDists <- dist( t( assay(rld) ) )
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste(rld$Genotype, rld$Stress, sep="-")
colnames(sampleDistMatrix) <- paste(rld$Genotype, rld$Stress, sep="-")
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

### Dendrogram of samples
```{r}
# Showing strain & media of each sample. Hierarchical clustering using rlog transformation.
options(repr.plot.width = 9, repr.plot.height = 5)
plot(hclust(sampleDists, method = "complete")) 
```

### Principal Components Analysis
```{r}
# Used to emphasize variation and bring out strong patterns in a dataset (dimensionality reduction).
# If two samples have similar levels of expression for the genes that contribute significantly to the variation represented by PC1, they will be plotted close together on the PC1 axis. 
plotPCA(rld, intgroup = c("Genotype"))
plotPCA(rld, intgroup = c("Stress"))
plotPCA(rld, intgroup = c("Genotype", "Stress"))
```

### The effect of Stress on the Genotype effect (the difference of difference) - See "Multi-factor designs" (http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs)
### Result 1 - OxsR vs H26 and Stress 
```{r}
ddsDE@design
resultsNames(ddsDE)

Res_Geno_Stress <- results(ddsDE, contrast=list("GenotypeOxsR.StressHOCl"))    # the difference (OxsR / H26) of difference (No / NaOCl)
Res_Geno_Stress
table_Geno_Stress <- Res_Geno_Stress %>%     # Make a result table
  data.frame() %>%
  rownames_to_column(var="Locus") %>%
  as_tibble()
table_Geno_Stress

table_Geno_Stress <- table_Geno_Stress[order(table_Geno_Stress$padj),]   # Sorted in ascending order by adjusted p-value
write.csv(table_Geno_Stress, file="Res1_OxsR_vs_H26_HOCl_vs_no.csv")

table_Geno_Stress_sig <- table_Geno_Stress %>%    # Threshold applied (significant genes)
  filter(abs(log2FoldChange) > 1 , padj < 0.05)
write.csv(table_Geno_Stress_sig, file="Res1_OxsR_vs_H26_HOCl_vs_no_sig.csv")
```

### Result 1 - Volcano plot
```{r}
#To visualize the data, "EnhancedVolcano" was used. 
#Blighe K, Rana S, Lewis M (2020). EnhancedVolcano: Publication-ready volcano plots with enhanced colouring and labeling. R package version 1.6.0, https://github.com/kevinblighe/EnhancedVolcano.

pdf(file = "Res1.pdf", width = 7, height = 7)

EnhancedVolcano(table_Geno_Stress,
                lab = rownames(table_Geno_Stress),
                labSize = 0,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-10, 10),
                title = 'Res1. OxsR_vs_H26 and HOCl_vs_no',
                xlab = bquote(~Log[2]~ 'fold change'),
                ylab = bquote(~-Log[10]~adjusted~italic(P)),
                pCutoff = 0.05,
                FCcutoff = 1,
                col = c('grey', 'grey', 'grey', 'red3'),
                colAlpha = 1)

dev.off()

#This code below was used to visualize the data with ggplot

#Time4h_Genotypetbs.vol.plot <- table_Time4h_GenotypetbsP %>% 
#  mutate(threshold_sig = padj < 0.05 & abs(log2FoldChange) > 1)

#ggplot(Time4h_Genotypetbs.vol.plot) +
#  geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold_sig)) +
#  ggtitle("Genotype (TbsP/pyrE) and Time (4h/30m) \n plus Glc") +
#  xlab("log2 fold change") + 
#  ylab("-log10 adjusted p-value") +
#  #scale_y_continuous(limits = c(0,50)) +
#  theme(legend.position = "none",
#        plot.title = element_text(size = rel(1.5), hjust = 0.5),
#        axis.title = element_text(size = rel(1.25)))  

```

### Genotype effect for the Time points
### Result 2 - OxsR vs H26 at the absent or present HOCl
```{r}
ddsDE$group <- factor(paste0(ddsDE$Stress, ddsDE$Genotype))
design(ddsDE) <- ~ group
dds.Genotype.effect <- DESeq(ddsDE)
resultsNames(dds.Genotype.effect)

summary(results(dds.Genotype.effect, contrast=c("group", "NoOxsR", "NoH26")))     # OxsR / H26 in the absence of HOCl (Res2-1)
summary(results(dds.Genotype.effect, contrast=c("group", "HOClOxsR", "HOClH26"))) # OxsR / H26 in the presence of HOCl (Res2-2)

# Result 2_1
Res_2_1 <- results(dds.Genotype.effect, contrast=c("group", "NoOxsR", "NoH26"))
table_Res_2_1 <- Res_2_1 %>%
  data.frame() %>%
  rownames_to_column(var="Locus") %>%
  as_tibble()

# Gen Key
anno.data <- read.csv("DS2_key.csv")
table_Res_2_1$Old_locus <- anno.data$Old_locus[match(table_Res_2_1$Locus, anno.data$Locus)]
table_Res_2_1$Protein_name <- anno.data$Protein_name[match(table_Res_2_1$Locus, anno.data$Locus)]

table_Res_2_1 <- table_Res_2_1[order(table_Res_2_1$padj),]
write.csv(table_Res_2_1, file="Res2_1_OxsR_vs_H26_No.csv")

table_Res_2_1_sig <- table_Res_2_1 %>%    # Threshold applied (significant genes)
  filter(abs(log2FoldChange) > 1 , padj < 0.05)
write.csv(table_Res_2_1_sig, file="Res2_1_OxsR_vs_H26_No_sig.csv")

# Result 2_2
Res_2_2 <- results(dds.Genotype.effect, contrast=c("group", "HOClOxsR", "HOClH26"))
table_Res_2_2 <- Res_2_2 %>%
  data.frame() %>%
  rownames_to_column(var="Locus") %>%
  as_tibble()

# Gen Key
#anno.data <- read.csv("DS2_uniprot_2.csv")
table_Res_2_2$Old_locus <- anno.data$Old_locus[match(table_Res_2_2$Locus, anno.data$Locus)]
table_Res_2_2$Protein_name <- anno.data$Protein_name[match(table_Res_2_2$Locus, anno.data$Locus)]

table_Res_2_2 <- table_Res_2_2[order(table_Res_2_2$padj),]
write.csv(table_Res_2_2, file="Res2_2_OxsR_vs_H26_HOCl.csv")

table_Res_2_2_sig <- table_Res_2_2 %>%    # Threshold applied (significant genes)
  filter(abs(log2FoldChange) > 1 , padj < 0.05)
write.csv(table_Res_2_2_sig, file="Res2_2_OxsR_vs_H26_HOCl_sig.csv")

```

### Result 2-1 - Volcano plot
```{r}
pdf(file = "Res2-1.pdf", width = 7, height = 7)

EnhancedVolcano(table_Res_2_1,
                lab = rownames(table_Res_2_1),
                labSize = 0,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-10, 10),
                title = 'Res2_1. OxsR_vs_H26 in the absence of HOCl',
                xlab = bquote(~Log[2]~ 'fold change'),
                ylab = bquote(~-Log[10]~adjusted~italic(P)),
                pCutoff = 0.05,
                FCcutoff = 1,
                col = c('grey', 'grey', 'grey', 'red3'),
                colAlpha = 1)

dev.off()
```

### Result 2-2 - Volcano plot
```{r}
pdf(file = "Res2-2.pdf", width = 7, height = 7)

EnhancedVolcano(table_Res_2_2,
                lab = rownames(table_Res_2_2),
                labSize = 0,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-10, 10),
                title = 'Res2_2. OxsR_vs_H26 in the presence of HOCl',
                xlab = bquote(~Log[2]~ 'fold change'),
                ylab = bquote(~-Log[10]~adjusted~italic(P)),
                pCutoff = 0.05,
                FCcutoff = 1,
                col = c('grey', 'grey', 'grey', 'red3'),
                colAlpha = 1)

dev.off()
```


### HOCl effect on each strain
### Result 3 - presence vs absnece in OxsR or H26
```{r}
ddsDE$group <- factor(paste0(ddsDE$Genotype, ddsDE$Stress))
design(ddsDE) <- ~ group
dds.Stress.effect <- DESeq(ddsDE)
resultsNames(dds.Stress.effect)

summary(results(dds.Stress.effect, contrast=c("group", "H26HOCl", "H26No")))    # HOCl / no on H26
summary(results(dds.Stress.effect, contrast=c("group", "OxsRHOCl", "OxsRNo")))   # HOCl / no on OxsR

# Result 3_1
Res_3_1 <- results(dds.Stress.effect, contrast=c("group", "H26HOCl", "H26No"))
table_Res_3_1 <- Res_3_1 %>%
  data.frame() %>%
  rownames_to_column(var="Locus") %>%
  as_tibble()

table_Res_3_1$Old_locus <- anno.data$Old_locus[match(table_Res_3_1$Locus, anno.data$Locus)]
table_Res_3_1$Protein_name <- anno.data$Protein_name[match(table_Res_3_1$Locus, anno.data$Locus)]

table_Res_3_1 <- table_Res_3_1[order(table_Res_3_1$padj),]
write.csv(table_Res_3_1, file="Res3_1_HCOl_vs_No_H26.csv")

table_Res_3_1_sig <- table_Res_3_1 %>%    # Threshold applied (significant genes)
  filter(abs(log2FoldChange) > 1 , padj < 0.05)
write.csv(table_Res_3_1_sig, file="Res3_1_HCOl_vs_No_H26_sig.csv")

#dim(table_Res_3_2_sig %>% filter(log2FoldChange<0))


# Result 3_2
Res_3_2 <- results(dds.Stress.effect, contrast=c("group", "OxsRHOCl", "OxsRNo"))
table_Res_3_2 <- Res_3_2 %>%
  data.frame() %>%
  rownames_to_column(var="Locus") %>%
  as_tibble()

table_Res_3_2$Old_locus <- anno.data$Old_locus[match(table_Res_3_2$Locus, anno.data$Locus)]
table_Res_3_2$Protein_name <- anno.data$Protein_name[match(table_Res_3_2$Locus, anno.data$Locus)]

table_Res_3_2 <- table_Res_3_2[order(table_Res_3_2$padj),]
write.csv(table_Res_3_2, file="Res3_2_HCOl_vs_No_OxsR.csv")

table_Res_3_2_sig <- table_Res_3_2 %>%    # Threshold applied (significant genes)
  filter(abs(log2FoldChange) > 1 , padj < 0.05)
write.csv(table_Res_3_2_sig, file="Res3_2_HCOl_vs_No_OxsR_sig.csv")

```

### Result 3-1 - Volcano plot
```{r}
pdf(file = "Res3-1.pdf", width = 7, height = 7)

EnhancedVolcano(table_Res_3_1,
                lab = rownames(table_Res_3_1),
                labSize = 0,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-10, 10),
                title = 'Res3_1. HCOl_vs_No in H26',
                xlab = bquote(~Log[2]~ 'fold change'),
                ylab = bquote(~-Log[10]~adjusted~italic(P)),
                pCutoff = 0.05,
                FCcutoff = 1,
                col = c('grey', 'grey', 'grey', 'red3'),
                colAlpha = 1)

dev.off()
```

### Result 3-2 - Volcano plot
```{r}
pdf(file = "Res3-2.pdf", width = 7, height = 7)

EnhancedVolcano(table_Res_3_2,
                lab = rownames(table_Res_3_2),
                labSize = 0,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-10, 10),
                title = 'Res3_2. HCOl_vs_No in OxsR',
                xlab = bquote(~Log[2]~ 'fold change'),
                ylab = bquote(~-Log[10]~adjusted~italic(P)),
                pCutoff = 0.05,
                FCcutoff = 1,
                col = c('grey', 'grey', 'grey', 'red3'),
                colAlpha = 1)

dev.off()
```

### Functional enrichment--code from Dulmage et al., 2015 (doi: 10.1128/mBio.00649-15)
# Load arCOG annotation and gene list files
```{r}
#rm(list = ls())
library(tidyverse)
library(dplyr)
library(ggplot2)

#Read in gene list file, in this case, a list of peaks significantly bound by OxsR
#genes <- read.csv ('Func_input_promoter_no.csv')  # NaOCl minus
genes <- read.csv ('Func_input_promoter_NaOCl_test.csv')  # NaOCl plus
cogs <- read.csv ("3_functional_enrichment/2014_2018_arCOG_HVO_renewed.csv")


## Prepare gene list (parse out gene unique IDs that are non-redundant and match those in cogfile)
genelist.p <- distinct(genes, locus_tag, .keep_all = TRUE)
genelist <- subset (genelist.p, select = 'locus_tag')
#genelist <- as.data.frame(genelist[-1, ]) #remove the first row - it's empty
#colnames(genelist) <- "locus_tag"

## Set up the hypergeometric test function
#Calculate statistical information about the cogs represented in the list of genes using the function below.
# Cogtest: three inputs to this function: 
# 1) list of gene names from the cluster (namelist)
# 2) COG file
# 3) p-value cutoff.  
#Note: can also do cogtest by different ID, just replace "cogfile$locus_tag" argument with other type of identifier, e.g. "cogfile$acc")
cogtest2=function(namelist, cogfile, pvalue) {
	cogs= subset(cogfile, is.element(cogfile$locus_tag, namelist$locus_tag)==TRUE)
	clust= summary(cogs$funclass_name)
	resm <- matrix(0, length(clust), 3)
	res <-data.frame(resm)
	rownames(res) <- names(clust)
	colnames(res) <- c("probability", "expect","count")
	all=summary(cogfile$funclass_name)
	totaln=sum(all)
	
		for (i in 1:length(clust)){
			res[i,1] <- phyper(clust[i], all[i], totaln-all[i], length(namelist$locus_tag),lower.tail=F)
			res[i,2] <- all[i]*(length(namelist$locus_tag)/totaln)
			res[i,3] <- clust[i]
		}
	subset(res,probability<=pvalue&count!=0)
}

## Use the following function to look at the genes in your cluster associated with a particular COG
#Note: the COGcategory name must be entered in quotes and add a space to the end of the category name of interest, e.g. 'transport '
cogset= function(namelist, cogfile, COGcategory) {
	subset(cogfile, is.element(cogfile$locus_tag, namelist$locus_tag)&is.element(cogfile$funclass_name, COGcategory)==TRUE)
}


## run the hypergeometric test to identify enriched functional categories
mycogs <- cogtest2(genelist, cogs, 1) # no p-value cutoff

all_mycogs <- mycogs %>%     # Make a result table
  data.frame() %>%
  rownames_to_column(var="COG_category") %>%
  as_tibble()
all_mycogs <- all_mycogs[order(all_mycogs$probability),]
write_csv(all_mycogs, "test.csv")

sig_mycogs <- all_mycogs %>%    # p-value cutoff applied
  filter(probability < 0.05)
write_csv(sig_mycogs, "test_sig.csv")  # _arCOG_sig

ggplot(sig_mycogs, aes(x=COG_category, y=count)) + 
  geom_bar(stat = "identity") + 
  geom_text(aes(label=count), vjust=0) +
  theme(axis.text.x=element_text(angle = -90, hjust=0)) + 
  theme(text = element_text(size=10))
```

### Dependency report
```{r}
sessionInfo()
```
